In [1]:
import warnings; warnings.simplefilter('ignore')

# libraries for data manipulation
import numpy as np 
import pandas as pd

In [2]:
# read csv dataset and load into pandas dataframe object "df"
df = pd.read_csv('complete.csv', encoding='utf_8')

In [3]:
df.head()

ID               name                     full_name  \
0   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
1  158023           L. Messi                  Lionel Messi   
2  190871             Neymar    Neymar da Silva Santos Jr.   
3  176580          L. Suárez                   Luis Suárez   
4  167495           M. Neuer                  Manuel Neuer   

                  club                                club_logo  special  age  \
0       Real Madrid CF  https://cdn.sofifa.org/18/teams/243.png     2228   32   
1         FC Barcelona  https://cdn.sofifa.org/18/teams/241.png     2158   30   
2  Paris Saint-Germain   https://cdn.sofifa.org/18/teams/73.png     2100   25   
3         FC Barcelona  https://cdn.sofifa.org/18/teams/241.png     2291   30   
4     FC Bayern Munich   https://cdn.sofifa.org/18/teams/21.png     1493   31   

                     league  birth_date  height_cm  ...  prefers_cb  \
0  Spanish Primera División  1985-02-05      185.0  ...       False   
1  Spanish Primera División  1987-06-24      170.0  ...       False   
2            French Ligue 1  1992-02-05      175.0  ...       False   
3  Spanish Primera División  1987-01-24      182.0  ...       False   
4         German Bundesliga  1986-03-27      193.0  ...       False   

  prefers_lb  prefers_lwb prefers_ls prefers_lf prefers_lam  prefers_lcm  \
0      False        False      False      False       False        False   
1      False        False      False      False       False        False   
2      False        False      False      False       False        False   
3      False        False      False      False       False        False   
4      False        False      False      False       False        False   

   prefers_ldm  prefers_lcb  prefers_gk  
0        False        False       False  
1        False        False       False  
2        False        False       False  
3        False        False       False  
4        False        False        True  

[5 rows x 185 columns]

In [4]:
df.columns

Index(['ID', 'name', 'full_name', 'club', 'club_logo', 'special', 'age',
       'league', 'birth_date', 'height_cm',
       ...
       'prefers_cb', 'prefers_lb', 'prefers_lwb', 'prefers_ls', 'prefers_lf',
       'prefers_lam', 'prefers_lcm', 'prefers_ldm', 'prefers_lcb',
       'prefers_gk'],
      dtype='object', length=185)

In [5]:
df

ID               name                     full_name  \
0       20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
1      158023           L. Messi                  Lionel Messi   
2      190871             Neymar    Neymar da Silva Santos Jr.   
3      176580          L. Suárez                   Luis Suárez   
4      167495           M. Neuer                  Manuel Neuer   
...       ...                ...                           ...   
17989  237463          A. Kelsey                   Adam Kelsey   
17990   11728      B. Richardson              Barry Richardson   
17991  231381           J. Young                  Jordan Young   
17992  238813       J. Lundstram                Josh Lundstram   
17993  238308          L. Sackey                 Leslie Sackey   

                      club                                 club_logo  special  \
0           Real Madrid CF   https://cdn.sofifa.org/18/teams/243.png     2228   
1             FC Barcelona   https://cdn.sofifa.org/18/teams/241.png     2158   
2      Paris Saint-Germain    https://cdn.sofifa.org/18/teams/73.png     2100   
3             FC Barcelona   https://cdn.sofifa.org/18/teams/241.png     2291   
4         FC Bayern Munich    https://cdn.sofifa.org/18/teams/21.png     1493   
...                    ...                                       ...      ...   
17989    Scunthorpe United  https://cdn.sofifa.org/18/teams/1949.png      755   
17990    Wycombe Wanderers  https://cdn.sofifa.org/18/teams/1933.png      832   
17991         Swindon Town  https://cdn.sofifa.org/18/teams/1934.png     1194   
17992      Crewe Alexandra   https://cdn.sofifa.org/18/teams/121.png     1302   
17993    Scunthorpe United  https://cdn.sofifa.org/18/teams/1949.png     1031   

       age                    league  birth_date  height_cm  ...  prefers_cb  \
0       32  Spanish Primera División  1985-02-05      185.0  ...       False   
1       30  Spanish Primera División  1987-06-24      170.0  ...       False   
2       25            French Ligue 1  1992-02-05      175.0  ...       False   
3       30  Spanish Primera División  1987-01-24      182.0  ...       False   
4       31         German Bundesliga  1986-03-27      193.0  ...       False   
...    ...                       ...         ...        ...  ...         ...   
17989   17        English League One  1999-11-12      188.0  ...       False   
17990   47        English League Two  1969-08-05      185.0  ...       False   
17991   17        English League Two  1999-07-31      175.0  ...       False   
17992   18        English League Two  1999-02-19      176.0  ...       False   
17993   18        English League One  1998-11-29      182.0  ...        True   

      prefers_lb  prefers_lwb prefers_ls prefers_lf prefers_lam  prefers_lcm  \
0          False        False      False      False       False        False   
1          False        False      False      False       False        False   
2          False        False      False      False       False        False   
3          False        False      False      False       False        False   
4          False        False      False      False       False        False   
...          ...          ...        ...        ...         ...          ...   
17989      False        False      False      False       False        False   
17990      False        False      False      False       False        False   
17991      False        False      False      False       False        False   
17992      False        False      False      False       False        False   
17993      False        False      False      False       False        False   

       prefers_ldm  prefers_lcb  prefers_gk  
0            False        False       False  
1            False        False       False  
2            False        False       False  
3            False        False       False  
4            False        False        True  
...            ...          ...         ...  
1

In [6]:
df.shape

(17994, 185)

In [7]:
#These means there are 17.994 players with 185 attributes

In [8]:
# Visualice the name of each columns and 
# the corresponding number for selection
for e, col in enumerate(df.columns):
    print(e,"->",col)

0 -> ID
1 -> name
2 -> full_name
3 -> club
4 -> club_logo
5 -> special
6 -> age
7 -> league
8 -> birth_date
9 -> height_cm
10 -> weight_kg
11 -> body_type
12 -> real_face
13 -> flag
14 -> nationality
15 -> photo
16 -> eur_value
17 -> eur_wage
18 -> eur_release_clause
19 -> overall
20 -> potential
21 -> pac
22 -> sho
23 -> pas
24 -> dri
25 -> def
26 -> phy
27 -> international_reputation
28 -> skill_moves
29 -> weak_foot
30 -> work_rate_att
31 -> work_rate_def
32 -> preferred_foot
33 -> crossing
34 -> finishing
35 -> heading_accuracy
36 -> short_passing
37 -> volleys
38 -> dribbling
39 -> curve
40 -> free_kick_accuracy
41 -> long_passing
42 -> ball_control
43 -> acceleration
44 -> sprint_speed
45 -> agility
46 -> reactions
47 -> balance
48 -> shot_power
49 -> jumping
50 -> stamina
51 -> strength
52 -> long_shots
53 -> aggression
54 -> interceptions
55 -> positioning
56 -> vision
57 -> penalties
58 -> composure
59 -> marking
60 -> standing_tackle
61 -> sliding_tackle
62 -> gk_diving
63 ->

In [9]:
# Dicard unnecesary rows
# There are 185 however the load of data
# will only be done with the selected columns

selectedColumns = [0,2,3,6,7,8,9,10,14,19]

# load data into a DataFrame object named 'df'
df = pd.read_csv('complete.csv', encoding='utf_8', usecols=selectedColumns)
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(df.head(5))

ID                     full_name                 club  age  \
0   20801  C. Ronaldo dos Santos Aveiro       Real Madrid CF   32   
1  158023                  Lionel Messi         FC Barcelona   30   
2  190871    Neymar da Silva Santos Jr.  Paris Saint-Germain   25   
3  176580                   Luis Suárez         FC Barcelona   30   
4  167495                  Manuel Neuer     FC Bayern Munich   31   

                     league  birth_date  height_cm  weight_kg nationality  \
0  Spanish Primera División  1985-02-05      185.0       80.0    Portugal   
1  Spanish Primera División  1987-06-24      170.0       72.0   Argentina   
2            French Ligue 1  1992-02-05      175.0       68.0      Brazil   
3  Spanish Primera División  1987-01-24      182.0       86.0     Uruguay   
4         German Bundesliga  1986-03-27      193.0       92.0     Germany   

   overall  
0       94  
1       93  
2       92  
3       92  
4       92

In [10]:
df.shape

(17994, 10)

In [11]:
df = df.set_index('ID')

In [12]:
df.head()

full_name                 club  age  \
ID                                                               
20801   C. Ronaldo dos Santos Aveiro       Real Madrid CF   32   
158023                  Lionel Messi         FC Barcelona   30   
190871    Neymar da Silva Santos Jr.  Paris Saint-Germain   25   
176580                   Luis Suárez         FC Barcelona   30   
167495                  Manuel Neuer     FC Bayern Munich   31   

                          league  birth_date  height_cm  weight_kg  \
ID                                                                   
20801   Spanish Primera División  1985-02-05      185.0       80.0   
158023  Spanish Primera División  1987-06-24      170.0       72.0   
190871            French Ligue 1  1992-02-05      175.0       68.0   
176580  Spanish Primera División  1987-01-24      182.0       86.0   
167495         German Bundesliga  1986-03-27      193.0       92.0   

       nationality  overall  
ID                           
20801     Portugal       94  
158023   Argentina       93  
190871      Brazil       92  
176580     Uruguay       92  
167495     Germany       92

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17994 entries, 20801 to 238308
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   full_name    17994 non-null  object 
 1   club         17741 non-null  object 
 2   age          17994 non-null  int64  
 3   league       17741 non-null  object 
 4   birth_date   17994 non-null  object 
 5   height_cm    17994 non-null  float64
 6   weight_kg    17994 non-null  float64
 7   nationality  17994 non-null  object 
 8   overall      17994 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 1.4+ MB


The next step is to prepare data for the analysis.

The preparation requieres:

4.1 Manage soccer players duplication (if any)
4.2 Manage empty values (if any)
4.3 Extract information from date of birth
4.4 Create categories for player's overall

In [14]:
# Look for duplicates
df.duplicated().sum()

0

In [15]:
# look for Nan values (if any)
pd.isnull(df).sum()

full_name        0
club           253
age              0
league         253
birth_date       0
height_cm        0
weight_kg        0
nationality      0
overall          0
dtype: int64

This shows that there are empty values in columns:

club

league

Each of these columns contain 253 empty values

In [16]:
# Display % clubs with NaN values
df.club.isnull().value_counts() / len(df)

False    0.98594
True     0.01406
Name: club, dtype: float64

In [17]:
# Display % leagues with NaN values
df.club.isnull().value_counts() / len(df)

False    0.98594
True     0.01406
Name: club, dtype: float64

Relevant facts:

Data missing is about 'club' and 'league',

Data missing is 1.5% from the total

I have decided to fill the Nan values of club and league to value "unknown" in order to proceed.

In [18]:
# club and league NaN values are asigned "Unknown"
df.club.fillna("Unknown", inplace=True)
df.league.fillna("Unknown", inplace=True)

In [19]:
# look for Nan values (if any)
pd.isnull(df).sum()

full_name      0
club           0
age            0
league         0
birth_date     0
height_cm      0
weight_kg      0
nationality    0
overall        0
dtype: int64

In [20]:
# Extraction of the month information from the 'birth_date' information
df['month'] = pd.DatetimeIndex(df['birth_date']).month
df['year'] = pd.DatetimeIndex(df['birth_date']).year

# Set the type of column 'month' and 'year' as int
df.month = df.month.astype(int)
df.year = df.year.astype(int)

In [21]:
#Lets create a column with the overall category of the player
df.overall.describe()

count    17994.000000
mean        66.253029
std          6.946729
min         46.000000
25%         62.000000
50%         66.000000
75%         71.000000
max         94.000000
Name: overall, dtype: float64

Looking at this we can see that the minimun value of overall for a player is

46

and the maximum value of overall is:

94

This information will be use to create different categories in scale of 10 units from 40 to 100

In [22]:
# Prepare a new column "overallCat" to save data about
# overall categories will be:
# 10, 20, 30, 40, 50, 60, 70, 80, 90, 100

# First we create and empty column:
df['overallCat'] = np.nan

# Then the column value is changed depending on the given conditions
df['overallCat'].loc[(df['overall'] > 40) & (df['overall'] <=50 )] = 40
df['overallCat'].loc[(df['overall'] > 50) & (df['overall'] <=60 )] = 50
df['overallCat'].loc[(df['overall'] > 60) & (df['overall'] <=70 )] = 60
df['overallCat'].loc[(df['overall'] > 70) & (df['overall'] <=80 )] = 70
df['overallCat'].loc[(df['overall'] > 80) & (df['overall'] <=90 )] = 80
df['overallCat'].loc[(df['overall'] > 90) & (df['overall'] <=100 )] = 90

# display the total values for each category
df.overallCat.value_counts()

60.0    9577
70.0    4434
50.0    3389
80.0     385
40.0     203
90.0       6
Name: overallCat, dtype: int64

In [23]:
df.shape

(17994, 12)

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17994 entries, 20801 to 238308
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   full_name    17994 non-null  object 
 1   club         17994 non-null  object 
 2   age          17994 non-null  int64  
 3   league       17994 non-null  object 
 4   birth_date   17994 non-null  object 
 5   height_cm    17994 non-null  float64
 6   weight_kg    17994 non-null  float64
 7   nationality  17994 non-null  object 
 8   overall      17994 non-null  int64  
 9   month        17994 non-null  int32  
 10  year         17994 non-null  int32  
 11  overallCat   17994 non-null  float64
dtypes: float64(3), int32(2), int64(2), object(5)
memory usage: 1.6+ MB


In [25]:
df

full_name                 club  age  \
ID                                                               
20801   C. Ronaldo dos Santos Aveiro       Real Madrid CF   32   
158023                  Lionel Messi         FC Barcelona   30   
190871    Neymar da Silva Santos Jr.  Paris Saint-Germain   25   
176580                   Luis Suárez         FC Barcelona   30   
167495                  Manuel Neuer     FC Bayern Munich   31   
...                              ...                  ...  ...   
237463                   Adam Kelsey    Scunthorpe United   17   
11728               Barry Richardson    Wycombe Wanderers   47   
231381                  Jordan Young         Swindon Town   17   
238813                Josh Lundstram      Crewe Alexandra   18   
238308                 Leslie Sackey    Scunthorpe United   18   

                          league  birth_date  height_cm  weight_kg  \
ID                                                                   
20801   Spanish Primera División  1985-02-05      185.0       80.0   
158023  Spanish Primera División  1987-06-24      170.0       72.0   
190871            French Ligue 1  1992-02-05      175.0       68.0   
176580  Spanish Primera División  1987-01-24      182.0       86.0   
167495         German Bundesliga  1986-03-27      193.0       92.0   
...                          ...         ...        ...        ...   
237463        English League One  1999-11-12      188.0       74.0   
11728         English League Two  1969-08-05      185.0       77.0   
231381        English League Two  1999-07-31      175.0       71.0   
238813        English League Two  1999-02-19      176.0       61.0   
238308        English League One  1998-11-29      182.0       72.0   

       nationality  overall  month  year  overallCat  
ID                                                    
20801     Portugal       94      2  1985        90.0  
158023   Argentina       93      6  1987        90.0  
190871      Brazil       92      2  1992        90.0  
176580     Uruguay       92      1  1987        90.0  
167495     Germany       92      3  1986        90.0  
...            ...      ...    ...   ...         ...  
237463     England       46     11  1999        40.0  
11728      England       46      8  1969        40.0  
231381    Scotland       46      7  1999        40.0  
238813     England       46      2  1999        40.0  
238308       Ghana       46     11  1998        40.0  

[17994 rows x 12 columns]

In [26]:
df.to_csv("collection.csv")

Error Management

During the analysis of the information on the dataset I have found out that, for some reason, all players in "Campeonato Brasileiro Série A" were born in february

I have looked into wikipedia for information regarding soccer players in the Campeonato Brasileiro Série A. It turns out that not all 320 players on this league were born in february.

Until I modify this dataset to contain the right information I have decided to delete all soccer players in "Campeonato Brasileiro Série A".

This is a total of 320 players to be erased

In [27]:
# lets select players form the league and count for each month how many there are. 

numPlayers = df[df['league']== "Campeonato Brasileiro Série A"]['month'].value_counts()
print("Number of players on Campeonato Brasileiro Série A by month of birth:\n",numPlayers)

Number of players on Campeonato Brasileiro Série A by month of birth:
 2    320
Name: month, dtype: int64


In [28]:
print("Before deleting there are: \t",(len(df.league.value_counts())), " leagues")

# Deletion of these players is done by:
df = df[df['league'] != 'Campeonato Brasileiro Série A']

print("After deleting there are: \t",(len(df.league.value_counts())), " leagues")

Before deleting there are: 	 42  leagues
After deleting there are: 	 41  leagues


In [30]:
import pandas as pd
import psycopg2
from sqlalchemy import create_engine
# Connect to the PostgreSQL database
conn = psycopg2.connect(
    host='localhost',
    port='5433',
    database='collection',
    user='postgres',
    password='Nit$0205'
)

df = pd.read_csv('collection.csv', encoding='utf_8')

engine = create_engine('postgresql+psycopg2://postgres:Nit$0205@localhost:5433/collection')

# Define the table name to update the DataFrame
table_name = 'football'

# Update the data in the PostgreSQL database
df.to_sql(table_name, engine, if_exists='replace', index=False)

# Close the database connection
conn.close()

print('Data updated in the PostgreSQL database.')

Data updated in the PostgreSQL database.
